src/h3/stream.rs

```diff
@@ -0,0 +1,516 @@
+pub enum StreamType {
+    Control,
+    Request,
+    Push,
+    QpackEncoder,
+    QpackDecoder,
+    // Grease, // TODO: enable GREASE streams
+}
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum StreamState {
+    StreamTypeLen,
+    StreamType,
+    FrameTypeLen,
+    FrameType,
+    FramePayloadLenLen,
+    FramePayloadLen,
+    FramePayload,
+    PushIdLen,
+    PushId,
+    QpackInstruction,
+    Invalid,
+    Done,
+}
+
+impl std::fmt::Debug for StreamState {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            StreamState::StreamTypeLen => {
+                write!(f, "StreamTypeLen")?;
+            },
+            StreamState::StreamType => {
+                write!(f, "StreamType")?;
+            },
+            StreamState::FrameTypeLen => {
+                write!(f, "FrameTypeLen")?;
+            },
+            StreamState::FrameType => {
+                write!(f, "FrameType")?;
+            },
+            StreamState::FramePayloadLenLen => {
+                write!(f, "FramePayloadLenLen")?;
+            },
+            StreamState::FramePayloadLen => {
+                write!(f, "FramePayloadLen")?;
+            },
+            StreamState::FramePayload => {
+                write!(f, "FramePayload")?;
+            },
+            StreamState::PushIdLen => {
+                write!(f, "PushIdLen")?;
+            },
+            StreamState::PushId => {
+                write!(f, "PushId")?;
+            },
+            StreamState::QpackInstruction => {
+                write!(f, "QpackInstruction")?;
+            },
+            StreamState::Invalid => {
+                write!(f, "Invalid")?;
+            },
+            StreamState::Done => {
+                write!(f, "Done")?;
+            },
+        }
+
+        Ok(())
+    }
+}
+
+impl StreamType {
+    // TODO: draft 18+ with require true varints
+    pub fn deserialize(v: u8) -> Option<StreamType> {
+
+        match v {
+            HTTP3_CONTROL_STREAM_TYPE_ID => Some(StreamType::Control),
+            HTTP3_PUSH_STREAM_TYPE_ID => Some(StreamType::Push),
+            QPACK_ENCODER_STREAM_TYPE_ID => Some(StreamType::QpackEncoder),
+            QPACK_DECODER_STREAM_TYPE_ID => Some(StreamType::QpackDecoder),
+            // TODO: parse grease stream
+            _ => {
+                trace!("Stream type value {:x} is unknown", v);
+                return None;
+            }
+        }
+    }
+
+    // TODO: draft 18+ with require true varints
+    pub fn _serialize(ty: StreamType) -> Option<u8> {
+        match ty {
+            StreamType::Control => Some(HTTP3_CONTROL_STREAM_TYPE_ID),
+            StreamType::Push => Some(HTTP3_PUSH_STREAM_TYPE_ID),
+            StreamType::QpackEncoder => Some(QPACK_ENCODER_STREAM_TYPE_ID),
+            StreamType::QpackDecoder => Some(QPACK_DECODER_STREAM_TYPE_ID),
+            _ => None,
+        }
+    }
+}
+
+impl std::fmt::Debug for StreamType {
```

Comment:
This also doesn't seem to be adding a whole lot of information compared to the derived `Debug`.